devtools/sendit/session.go

```diff
@@ -0,0 +1,211 @@
+package sendit
+
+import (
+	"context"
+	"errors"
+	"io"
+	"log"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/hashicorp/yamux"
+)
+
+type session struct {
+	ID     string
+	Prefix string
+
+	stream *Stream
+	ym     *yamux.Session
+	s      *Server
+
+	mx sync.Mutex
+
+	start sync.Once
+
+	reader chan *ioContext
+
+	doneCh  chan struct{}
+	readyCh chan struct{}
+
+	pending int32
+}
+
+func (s *Server) newSession(prefix string) (*session, error) {
+	if !ValidPath(prefix) {
+		return nil, errors.New("invalid prefix value")
+	}
+	id, err := genID()
+	if err != nil {
+		panic(err)
+	}
+
+	s.mx.Lock()
+	defer s.mx.Unlock()
+	if _, ok := s.sessionsByPrefix[prefix]; ok {
+		return nil, errors.New("prefix already in use")
+	}
+
+	sess := &session{
+		ID:     id,
+		Prefix: prefix,
+		s:      s,
+
+		stream:  NewStream(),
+		reader:  make(chan *ioContext, 1),
+		doneCh:  make(chan struct{}),
+		readyCh: make(chan struct{}),
+	}
+
+	s.sessionsByPrefix[prefix] = sess
+	s.sessionsByID[id] = sess
+
+	return sess, nil
+}
+
+func (sess *session) init() {
+	select {
+	case <-sess.doneCh:
+		return
+	default:
+	}
+
+	select {
+	case <-sess.stream.Ready():
+	case <-sess.doneCh:
+		return
+	}
```

Comment:
do we need to be repeating this logic waiting on the done channel to return?

Reply:
Yes, `select` makes no guarantees on order and we want to exit if `doneCh` is set, even if the stream is ready. If we only had one, it would be effectively random which one would get picked each time, even if closed.

Reply:
Can we document that here and other relevant places for readability and maintenance?